File: 2/js/src/LevinshteinDistance.ts
function solve(source: string, target: string): number {
	const n = source.length;
	const m = target.length;

	let edits: number[][] = [];

	for(let i=0; i<=n; i++) {
		edits.push([]);
	}

	for(let i=0; i<=m; i++) {
		edits[0][i] = i;
	}

	for(let i=0; i<=n; i++) {
		edits[i][0] = i;
	}

	for(let i=1; i<=n; i++) {
		for(let j=1; j<=m; j++) {
			if(source.charAt(i-1) == target.charAt(j-1)) {
				edits[i][j] = edits[i-1][j-1];
			} else {
				edits[i][j] = Math.min(edits[i-1][j], edits[i][j-1], edits[i-1][j-1]) + 1;
			}
		}
	}

	return edits[n][m];
}

(function main() {
	const source = "abc";
	const target = "yabd";

	console.log(solve(source, target));
})();

File: 2/java/LevinshteinDistance.java
class LevinshteinDistance {
	public static int getLevinshteinDistance(String str1, String str2) {
		int n = str1.length();
		int m = str2.length();
		int matrix[][] = new int[n+1][m+1];	
		
		for(int i=0; i<=m; i++){
			matrix[0][i] = i;
		}

		for(int i=1; i<=n; i++){
			matrix[i][0] = i;
		}
	
		for(int r=1; r<=n; r++) {
			for(int c=1; c<=m; c++) {
				if(str1.charAt(r-1) == str2.charAt(c-1)) {
					matrix[r][c] = matrix[r-1][c-1];
				} else {
					matrix[r][c] = Math.min(matrix[r-1][c-1], Math.min(matrix[r-1][c], matrix[r][c-1])) + 1;
				}
			}
		}

		return matrix[n][m];
	}

	public static void main(String args[]) {
		String str1 = "abc";
		String str2 = "yabd";

		System.out.println(getLevinshteinDistance(str1, str2));
	}
}
